File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer,ArrayList<Integer>> mp=new HashMap<>();

        for(int i = 0; i < n; i++){
            mp.put(i, new ArrayList<>());
        }

        for(int i=0;i<edges.length;i++){
            int a = edges[i][0];
            int b = edges[i][1];

            mp.get(a).add(b);
            mp.get(b).add(a); // FIX 1: undirected graph
        }

        int visited[]=new int[n];
        return solve(mp,source,destination,visited); 
    }

    public boolean solve(Map<Integer,ArrayList<Integer>> mp, int source, int destination, int visited[]){
        if(source==destination){
            return true;
        }

        visited[source]=1;

        for(int next : mp.get(source)){
            if(visited[next]==0){
                boolean ans = solve(mp,next,destination,visited);
                if(ans){
                    return true;
                }
            }
        }

        return false;
    }
}

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = defaultdict(list)
        for i in range(n):
            graph[i] = []

        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)   
        visited = [0] * n
        return self.dfs(graph, source, destination, visited)
    def dfs(self, graph, source, destination, visited):
        if source == destination:
            return True

        visited[source] = 1

        for nxt in graph[source]:
            if visited[nxt] == 0:
                if self.dfs(graph, nxt, destination, visited):
                    return True

        return False

        
        